Add function for composing complex mapped pipelines

jcrist opened the pull request:
Currently mapping in prefect is commonly done using the `task.map` api. This works well for simple pipelines, but fails for cases where you want to embed complicated patterns inside the mapped pipelines (e.g. control flow). Some work to remedy this was done in #2784, but this doesn't work for `case` blocks.

This provides an alternative solution. A new method `compose_map` (there's likely a better name) is added. This takes a function that adds multiple tasks operating on scalar arguments, and maps it across one or more iterable task inputs. This is best explained with an example:

```python
from prefect import task, case, Flow
from prefect.utilities.tasks import compose_map
from prefect.tasks.control_flow import merge


@task
def add(x, y):
    return x + y


@task
def is_even(x):
    return x % 2 == 0


def inc_if_even(x):
    with case(is_even(x), True):
        res = add(x, 1)
    return merge(res, x)


with Flow("test") as flow:
    res = compose_map(inc_if_even, range(10))
```

This generates the following flow structure:

![example](https://user-images.githubusercontent.com/2783717/85469348-2bdd2080-b573-11ea-83f6-c4046b1ff16d.png)

Like `task.map`, `unmapped` arguments can be passed as `args` or `kwargs` to avoid being mapped. It also works with the imperative api - a `flow` object can be explicitly passed.

To prevent misuse, an error is raised if the user tries to map anything from inside the function passed to `compose_map`, since prefect doesn't support mapped tasks inside mapped tasks. The error is implemented at the imperative api level, so it raises appropriately for all cases.

- [x] adds new tests (if appropriate)
- [x] add a changelog entry in the `changes/` directory (if appropriate)
- [x] updates docstrings for any new functions or function arguments, including `docs/outline.toml` for API reference docs (if appropriate)

jlowin commented:
@jcrist 👍 great way to bridge the gap between naive composition and a Prefect-native solution (without resorting to subflows and all the complexity they entail for users)!

After playing with it for a bit, my updated name suggestion is `apply_map` or `as_map`. Not sure if that's final yet, but when using it in code, the feeling I get is "apply this factory function to these arguments" just as an `apply*` function might work in pandas.

jcrist commented:
Two open questions:

1. Should this be a top-level function? I think it should be, but could go either way.

2. Should all tasks created under an `apply_map` call have all args to `apply_map` set as upstream tasks? This is both a semantic question and an implementation question. Two examples:

```python
@task
def inc(x):
    return x + 1

def func(a, b):
    x = inc(a)
    y = inc(b)
    return x, y

with Flow("test"):
    a = range(3)
    b = range(3, 6)
    x, y = apply_map(func, a, b)
```

Given the above, `x` and `y` have no shared upstream tasks - this means that if `a` succeeds and `b` fails, `x` would run and `b` would not. *Also, since they have no shared upstream tasks, they're functionally independent mapping operations - `a` and `b` could have different lengths!* `a` could have 3 elements while `b` could have 100, and the `apply_map` call (and flow) would succeed.

```python
@task
def add(x, y):
    return x + y

def func(a):
    c = inc(1)
    return add(a, c)

with Flow("test") as flow:
    apply_map(func, range(3))
```

In this case `c` has no *external* upstream tasks (the tree of upstream tasks ends with constants). With no recursive external dependencies, there's nothing making the `inc` task run as a mapped task at all, but semantically it should be. To get around this now, any task with no upstream tasks rooted in the arguments to `apply_map` gets all the non-unmapped arguments set as upstream tasks.

![example](https://user-images.githubusercontent.com/2783717/85776010-eb51e400-b6e5-11ea-8b08-90dd007fe166.png)

This means that like everything else added in `func`, `inc(1)` runs once for every mapped pipeline.

This works, but the generated graphs may be confusing for users. I think this is likely an edge case, but it probably will come up.

I think the root of this question comes down to "are all tasks created with `apply_map` semantically a unit of work". If the examples above were running eagerly like normal Python code, `apply_map` wouldn't be called until all its arguments were ready, and wouldn't return until all of its outputs were completed. As written, depending on the internal dependency structure this is more like a macro - the graph created by `apply_map` may have separate components with separate dependency structures, there's no `input -> tasks-created-by-apply-map -> output` flow.

I find the existing behavior of only adding the args of `apply_map` as upstream dependencies of dependency-less tasks to be a bit confusing during edge cases (it might result in weird behavior for users doing complicated things) - but under normal operation things "just work".

One uniform option would be to enforce that the full sub-graph created under `apply_map` shares upstream dependencies with all of its args. We do this with `case` already - any task created under a case statement has the case statement as a direct or indirect dependency. This would mean that in the first example:

- `a` and `b` would be dependencies of both `x` and `y`.
- `a` and `b` must have the same lengths.
- If `a` fails, `y` would also fail, even though `a` isn't a data dependency of `y` (but still is an upstream task).

After writing this all out, this is what I'm leaning towards, but still want a second opinion.

jcrist commented:
> One uniform option would be to enforce that the full sub-graph created under apply_map shares upstream dependencies with all of its args. 

Ended up going with this. All tasks created inside `apply_map` will have all non unmapped-constant (i.e. not `unmapped(1)`) arguments to `apply_map` set as transitive upstream dependencies. This ensures both that all args to `apply_map` have the same mapped length, and that failure states upstream propagate properly through child tasks created in `apply_map` as a whole.

I've also added `apply_map` as a top-level function, and added it to the mapping section of the docs.

I believe this is ready for review.

jlowin commented:
I think having `apply_map` as a top-level function is appropriate given the other top-level functions, but as that top-level gets more crowded we may want to codify what earns something a place there. Loosely, to me it's been "a central part of the Python syntax": key objects (`task`, `Flow`), context managers (`context`, `tags`, `case`) and modifiers (`unmapped`, `apply_map`). Perhaps that should be codified, or these should be moved into submodules for organization and we should have a new "recommended" global import pattern. 

I'm using your thought as an excuse to riff on this but this is certainly not intended to hold up your PR.

jcrist commented:
I think things that are general (core objects, context/config/tags, and control flow things like `case`/`apply_map`) make sense (the top level namespace has things likely to be used in any flows). Anything domain specific should be a more qualified import. I also agree, we do need to strike a balance between ease of import and crowding the top-level namespace.

Either way, this doesn't apply here, since we both think `apply_map` does belong top-level.

cicdw commented on src/prefect/utilities/tasks.py:
```diff
@@ -8,13 +10,170 @@
+    # Used to convert constants to constant tasks if needed
+    id_to_const = {}
+
+    def preprocess(a: Any) -> "prefect.Task":
```
Could you add a comment to this block (L95-L107) that this is where the inputs to `apply_map` are added to the flow?

jcrist replied:
Done.

jcrist commented on src/prefect/utilities/tasks.py:
```diff
@@ -8,13 +10,170 @@
+            if no_flow_provided:
+                res = func(*args2, **kwargs2)
+            else:
+                res = func(*args2, flow=flow2, **kwargs2)
```
See the docstring on `flow` here: https://github.com/PrefectHQ/prefect/blob/3e919c399e7cf5baa149cc6cdf52371504493f41/src/prefect/utilities/tasks.py#L42-L44

If `flow` is provided as a parameter, we assume the user is using the imperative api inside `func` (and that there's no active `flow` in context). This feels intuitive to me at least, since prefect has a convention in other places of passing in `flow` if one isn't/can't be inferred.

cicdw replied:
Yea that all makes sense to me, but this code block is passing a `Flow` to the user-provided `func` which doesn't seem necessary; whether the user is using the functional or imperative API, your `with flow2:` takes care of putting the relevant Flow object in context, so you could have:
```python
with flow2:
    res = func(*args2, **kwargs2)
```

To be more concrete, this errors out:
```python
from prefect import task, Flow, apply_map

@task
def random_task(y):
    return y + 1

def mapper(x):
    z = random_task(x)
    q = random_task(z)

f = Flow("example")

apply_map(mapper, range(10), flow=f)
# TypeError: mapper() got an unexpected keyword argument 'flow'
```

jcrist replied:
Right, but that's documented in the docstring to `apply_map`.

> If specified, `func` must accept a `flow` keyword argument. 

My main goal here was to make this easy to use with both the functional and imperative api. If a user wants to map a function like:

```python
def func(x, flow):
    flow.add_task(foo)
    flow.add_edge(upstream_task=bar, downstream_task=foo)

apply_map(func, x)
```

should they grab the flow from the context instead, or wrap it with a lambda/partial to ensure the `flow` is passed in? I can see arguments for either behavior - the current design (if you pass in a `flow`, it's forwarded to the `func`) to me doesn't feel *wrong*, just different than what you're expecting. Happy to change, just want to clarify why things are the way they are.

cicdw replied:
Ahhhh ok ok I missed that in the docstring, I understand now.  Yea I think that's a great callout, as there are some users that live entirely within the imperative API.   I'm 👍 on this now, thanks for explaining

jcrist commented:
> One other question: is there any risk that an error gets raised midway through the logic that leaves the original flow in a half-altered state? I definitely do not want to over engineer a solution for this, but if it's something that could happen easily I'd want to at least consider it

I don't think so, at least not one that they could recover from using normal patterns. We pre-check all args to `apply_map` to ensure they're valid before adding any of them to the flow, so the only invalid state I can think of is a cycle in the graph (unlikely in normal functional-only flow construction, since you can't reference a task as an argument that hasn't been created yet). I think the cost of checking that isn't worth the benefit - the user will get an error later on in that case, and it's not clear what the error would be here, or how they could recover from it.
